lib/gameState/GameStatistics.cpp

```diff
@@ -0,0 +1,328 @@
+	}
+	return str;
+}
+
+// get total gold income
+int Statistic::getIncome(const CGameState * gs, const PlayerState * ps)
+{
+	int percentIncome = gs->getStartInfo()->getIthPlayersSettings(ps->color).handicap.percentIncome;
+	int totalIncome = 0;
+	const CGObjectInstance * heroOrTown = nullptr;
+
+	//Heroes can produce gold as well - skill, specialty or arts
+	for(const auto & h : ps->heroes)
+	{
+		totalIncome += h->valOfBonuses(Selector::typeSubtype(BonusType::GENERATE_RESOURCE, BonusSubtypeID(GameResID(GameResID::GOLD)))) * percentIncome / 100;
+
+		if(!heroOrTown)
+			heroOrTown = h;
+	}
+
+	//Add town income of all towns
+	for(const auto & t : ps->towns)
+	{
+		totalIncome += t->dailyIncome()[EGameResID::GOLD];
+
+		if(!heroOrTown)
+			heroOrTown = t;
+	}
+
+	for(const CGMine * mine : getMines(gs, ps))
+	{
+		if (mine->producedResource == EGameResID::GOLD)
+			totalIncome += mine->getProducedQuantity();
+	}
+
+	return totalIncome;
+}
+
+double Statistic::getMapVisitedRatio(const CGameState * gs, PlayerColor player)
+{
+	double visible = 0.0;
+	double numTiles = 0.0;
+
+	for(int layer = 0; layer < (gs->map->twoLevel ? 2 : 1); layer++)
+		for(int y = 0; y < gs->map->height; ++y)
+			for(int x = 0; x < gs->map->width; ++x)
+			{
+				TerrainTile tile = gs->map->getTile(int3(x, y, layer));
+
+				if(tile.blocked && (!tile.visitable))
+					continue;
+
+				if(gs->isVisible(int3(x, y, layer), player))
+					visible++;
+				numTiles++;
+			}
+	
+	return visible / numTiles;
+}
+
+const CGHeroInstance * Statistic::findBestHero(const CGameState * gs, const PlayerColor & color)
+{
+	auto &h = gs->players.at(color).heroes;
+	if(h.empty())
+		return nullptr;
+	//best hero will be that with highest exp
+	int best = 0;
+	for(int b=1; b<h.size(); ++b)
+	{
+		if(h[b]->exp > h[best]->exp)
+		{
+			best = b;
+		}
+	}
+	return h[best];
+}
+
+std::vector<std::vector<PlayerColor>> Statistic::getRank(std::vector<TStat> stats)
+{
+	std::sort(stats.begin(), stats.end(), [](const TStat & a, const TStat & b) { return a.second > b.second; });
+
+	//put first element
+	std::vector< std::vector<PlayerColor> > ret;
+	std::vector<PlayerColor> tmp;
+	tmp.push_back( stats[0].first );
+	ret.push_back( tmp );
+
+	//the rest of elements
+	for(int g=1; g<stats.size(); ++g)
+	{
+		if(stats[g].second == stats[g-1].second)
+		{
+			(ret.end()-1)->push_back( stats[g].first );
+		}
+		else
+		{
+			//create next occupied rank
+			std::vector<PlayerColor> tmp;
+			tmp.push_back(stats[g].first);
+			ret.push_back(tmp);
+		}
+	}
+
+	return ret;
+}
+
+int Statistic::getObeliskVisited(const CGameState * gs, const TeamID & t)
+{
+	if(gs->map->obelisksVisited.count(t))
+		return gs->map->obelisksVisited.at(t);
+	else
+		return 0;
+}
+
+double Statistic::getMightMagicRatio(const PlayerState * ps)
+{
+	double numMight = 0;
+
+	for(auto h : ps->heroes)
+		if(h->type->heroClass->affinity == CHeroClass::EClassAffinity::MIGHT)
+			numMight++;
+
+	return numMight / ps->heroes.size();
+}
```

Comment:
Not sure if this stat is 'interesting' in any way. Especially since it does not accounts for hero strength.
If I am playing with Crag Hack who gets all army & exp and bunch of lvl 1 magic heroes-scouts my 'might magic ratio' would lean to magic.

Might be easier to remove it probably.
Asking pvp players for what stats they might be interested will likely give you better ideas than this one.

Reply:
Yes, I've noticed that too. :)
Removed